File: geeksforgeeks/Algorithms/Searching and Sorting/merge_sort.java
import java.util.*;
public class Main {
	static void merge(int arr[], int l, int m, int r) //Merge Function
	{
	    int i, j, k;
	    int n1 = m - l + 1;
	    int n2 =  r - m;
	 
	    int L[] = new int[n1];
	    int R[] = new int[n2];
	 
	    for(i = 0; i < n1; i++)
	        L[i] = arr[l + i];
	    for(j = 0; j < n2; j++)
	        R[j] = arr[m + 1+ j];
	 
	    i = 0;
	    j = 0;
	    k = l;
	    while (i < n1 && j < n2)
	    {
	        if (L[i] <= R[j])
	        {
	            arr[k] = L[i];
	            i++;
	        }
	        else
	        {
	            arr[k] = R[j];
	            j++;
	        }
	        k++;
	    }
	 
	    while (i < n1)
	    {
	        arr[k] = L[i];
	        i++;
	        k++;
	    }
	 
	    while (j < n2)
	    {
	        arr[k] = R[j];
	        j++;
	        k++;
	    }
	}
	
	static void mergeSort(int arr[],int l,int r) //Merge Sort Function
	{
		if(r>l)
		{
			int mid =(l+r)/2;
			
			mergeSort(arr,l,mid);
			mergeSort(arr,mid+1,r);
			merge(arr, l, mid, r);
		}
	}
	
	public static void main(String[] args) 
	{
		int arr[]={12,11,5,3,6,343,2,43,1,232,45};
		int n = arr.length;
		
		mergeSort(arr,0,n-1);
		
		for(int i=0;i<n;i++)
		{
			System.out.println(arr[i]);
		}
		
	}  
}

File: geeksforgeeks/Top 10 Algorithms and Data Structures/Searching and Sorting/Merge Sort.cpp
#include <iostream>
using namespace std; 

// The merge function
void merge(int arr[], int start, int mid, int end)
{
    int n1 = mid - start + 1;
    int n2 = end - mid;
    int L[n1],R[n2];
    
    for(int i=0;i < n1; i++)
    {
        L[i] = arr[start + i];
    }
    
    for(int i=0;i < n2; i++)
    {
        R[i] = arr[mid + 1 + i];
    }
    
    int i,j,l;
    i=j=0;
    l = start;
    
    while(i<n1 && j<n2)
    {
        if(L[i]<=R[j])
        {
            arr[l] = L[i];
            i++;
        }
        else
        {
            arr[l] = R[j];
            j++;
        }
        l++;
    }
    
    while(i<n1)
    {
        arr[l] = L[i];
        i++;
        l++;
    }
    
    while(j<n2)
    {
        arr[l] = R[j];
        j++;
        l++;
    }
}

// The mergeSort function
void mergeSort(int arr[], int start, int end)
{
    if(start < end)
    {
        int mid = (start + end)/2;
        mergeSort(arr,start,mid);
        mergeSort(arr,mid+1,end);
        merge(arr,start,mid,end);
    }
}

int main() {
    // Test Array
    int arr[] = {3,2,43,5,65,1,23,2,43,5,1};
    int end = sizeof(arr)/sizeof(arr[0]);
    
    mergeSort(arr,0,end-1);
    
    // Sorted Array
    for(int i=0;i<end; i++)
    {
        cout << arr[i] << " ";
    }
    return 0;
}